src/app/code/community/Omise/Gateway/Model/Payment/Base/Payment.php

```diff
@@ -0,0 +1,193 @@
+    {
+        $this->_infoBlockType = $infoBlockType;
+    }
+
+    /**
+     * @param bool $isGateway
+     */
+    public function setIsGateway($isGateway)
+    {
+        $this->_isGateway = $isGateway;
+    }
+
+    /**
+     * @param bool $canReviewPayment
+     */
+    public function setCanReviewPayment($canReviewPayment)
+    {
+        $this->_canReviewPayment = $canReviewPayment;
+    }
+
+    /**
+     * @param bool $isInitializeNeeded
+     */
+    public function setIsInitializeNeeded($isInitializeNeeded)
+    {
+        $this->_isInitializeNeeded = $isInitializeNeeded;
+    }
+
+    /**
+     * @param string $callbackUrl
+     */
+    public function setCallbackUrl($callbackUrl)
+    {
+        $this->_callbackUrl = $callbackUrl;
+    }
+
+    /**
+     * @return string
+     */
+    public function getCallbackUrl()
+    {
+        return $this->_callbackUrl;
+    }
+
+    public function __construct()
+    {
+        parent::__construct();
+    }
+
+    /**
+     * Instantiate state and set it to state object
+     *
+     * @param string        $payment_action
+     * @param Varien_Object $state_object
+     */
+    public function initialize($payment_action, $state_object)
+    {
+        $title =  (string)Mage::getConfig()->getNode('default/payment/'.$this->getCode().'/title');
+        $payment = $this->getInfoInstance();
+        $order   = $payment->getOrder();
+
+        $invoice = $order->prepareInvoice();
+        $invoice->setIsPaid(false)->register();
+
+        $charge = $this->process($payment, $invoice->getBaseGrandTotal());
+
+        $payment->setCreatedInvoice($invoice)
+            ->setIsTransactionClosed(false)
+            ->setIsTransactionPending(true)
+            ->addTransaction(
+                Mage_Sales_Model_Order_Payment_Transaction::TYPE_ORDER,
+                $invoice,
+                false,
+                Mage::helper('omise_gateway')->__('Processing an amount of %s via Omise '.$title.' payment.', $order->getBaseCurrency()->formatTxt($invoice->getBaseGrandTotal()))
+            );
+
+        $order->addRelatedObject($invoice);
+
+        if ($charge->isAwaitPayment()) {
+            $state_object->setState(Mage_Sales_Model_Order::STATE_PENDING_PAYMENT);
+            $state_object->setStatus($order->getConfig()->getStateDefaultStatus(Mage_Sales_Model_Order::STATE_PENDING_PAYMENT));
+            $state_object->setIsNotified(false);
+
+            return;
+        }
+
+        $this->_suspectToBeFailed($payment);
+    }
+
+    /**
+     * Check method for processing with base currency
+     * Note that Internet Banking can only be used with Omise Thailand account and 'THB' currency
+     *
+     * @param  string $currencyCode
+     *
+     * @return boolean
+     */
+    public function canUseForCurrency($currencyCode)
+    {
+        return (strtoupper($currencyCode) === 'THB' && strtoupper(Mage::app()->getStore()->getCurrentCurrencyCode()) === 'THB');
```

Comment:
Should probably make this look at a class variable (possibly an array) that defines the currency or currencies that a particular payment method is available for. This would make this generic `canUseForCurrency` method more flexible and remove the need to override it in as many cases